# Clasificación Multi-Clase.

Como el nombre indica, es un método para clasificar datasets con más de 2 clases.

In [1]:
%matplotlib notebook
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split

#Tomamos el dataset que usaremos:
fruits=pd.read_table('fruit_data_with_colors.txt')

X_fruits_2d=fruits[['height', 'width']]
y_fruits_2d=fruits['fruit_label']

C:\Users\jlxdp\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\t'.
  # Remove the CWD from sys.path while we load stuff.


In [2]:
from adspy_shared_utilities import plot_class_regions_for_classifier_subplot
from sklearn.svm import LinearSVC

X_train,X_test,y_train,y_test=train_test_split(X_fruits_2d,y_fruits_2d)

svm=LinearSVC(C=5,random_state=67)
svm.fit(X_train,y_train)

C:\Users\jlxdp\Anaconda3\lib\site-packages\sklearn\svm\base.py:931: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


LinearSVC(C=5, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=67, tol=0.0001,
     verbose=0)

Veamos el resultado.

In [3]:
print(svm.coef_)
print(svm.intercept_)

[[-0.30470481  0.51284339]
 [-1.63356808  1.14598843]
 [-0.08705913  0.37564613]
 [ 1.68994343 -2.03114177]]
[-1.96946116  1.20273085 -2.74817389  0.30410929]


Analizemos un ejemplos, en este caso de la clase 1 apple:<br>
\begin{equation*}
    y_{apple}=-0.3*height+0.51*width-1.97
\end{equation*}<br>
h=2. w=6.<br>
\begin{equation*}
    y_{apple}=-0.3*2+0.51*6-1.97\;(>=0: apple)
\end{equation*}<br>
h=2. w=2.<br>
\begin{equation*}
    y_{apple}=-0.3*2+0.51*2-1.97\;(<0: not\;an\;apple)
\end{equation*}<br>